### ISSUE TITLE
Refactor multidms to use jaxmodels as core modeling engine

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#169]
# Refactor multidms to use jaxmodels as core modeling engine

## Overview

This issue outlines a comprehensive refactoring plan to integrate the new `jaxmodels` module (PR #164) as the core modeling engine for multidms, replacing the current implementation in `multidms.model` and `multidms.biophysical`. The new architecture provides cleaner separation of concerns, improved numerical stability, and better performance.

## Background

The `jaxmodels` module introduces several key improvements:
- **Cleaner architecture** using Equinox modules with clear separation between data, models, and optimization
- **Improved parameterization** eliminating redundancy with a standard fitness-based approach
- **Better conditioning** using simpler preconditioning instead of expensive ADMM
- **Count-based loss** option for direct modeling of count data with negative binomial
- **Removed dependencies** (e.g., pandarallel)

#### 1.1 Create adapter layer for backward compatibility
- [ ] Create `multidms.model_v2.Model` class that wraps `jaxmodels.Model`
- [ ] Implement compatibility layer in `multidms.Data` to support both old and new data formats
- [ ] Add deprecation warnings for old API usage

#### 1.2 Refactor data handling
- [ ] Update `multidms.Data` to optionally use `jaxmodels.Data` internally
- [ ] Add method `Data.to_jaxmodels()` to convert to new format
- [ ] Maintain backward compatibility for existing data loading

#### 1.3 Replace core modeling components
- [ ] Replace optimization logic in `Model.fit()` with `jaxmodels.fit()`
- [ ] Map existing parameters to new parameterization:
  - `scale_coeff_lasso_shift` → `fusionreg`
  - `beta` + `shift` → separate `β` per condition
  - `theta_scale/bias` → handled by global epistasis
- [ ] Update parameter initialization to use new warmstart approach

#### 2.1 Global epistasis models
- [ ] Map existing epistasis functions to new architecture:
  - `sigmoidal_global_epistasis` → `jaxmodels.Sigmoid`
  - `identity_activation` → `jaxmodels.Identity`
  - Add softplus, NN variants as needed
- [ ] Ensure all existing epistasis options work with new system

#### 2.2 Loss functions
- [ ] Support both functional score loss (Huber) and count-based loss
- [ ] Add configuration option to select loss type
- [ ] Maintain backward compatibility with existing loss parameters

#### 2.3 Model fitting options
- [ ] Map existing fit parameters to new system:
  - `tol`, `maxiter`, `maxls` → `ge_kwargs`, `cal_kwargs`
  - `acceleration` → handled by optimizer
  - `lock_params` → needs new implementation
  - `convergence_trajectory_resolution` → loss trajectory

#### 3.1 Preserve existing analysis methods
- [ ] Update `Model.get_mutations_df()` to work with new parameter structure
- [ ] Ensure `Model.phenotype_as_effect_type_df()` compatibility
- [ ] Update cross-validation methods for new architecture

#### 3.2 Update plotting functionality
- [ ] Refactor plotting methods to work with new model structure
- [ ] Ensure all existing plots remain functional
- [ ] Consider moving plots to separate module for cleaner separation

#### 3.3 ModelCollection compatibility
- [ ] Update `ModelCollection` to work with new Model class
- [ ] Ensure `fit_models()` function works with new architecture
- [ ] Update parallel fitting to use new fit function

#### 4.1 Migration guide
- [ ] Create migration guide for users
- [ ] Document parameter mapping between old and new APIs
- [ ] Provide code examples for common use cases

#### 4.2 Update documentation
- [ ] Update docstrings for all modified classes/functions
- [ ] Update notebooks to use new API (with compatibility notes)
- [ ] Update biophysical model documentation

#### 4.3 Testing
- [ ] Create comprehensive test suite comparing old vs new implementations
- [ ] Run `simulation_validation.ipynb` with new code
- [ ] Benchmark performance improvements
- [ ] Validate numerical accuracy matches or improves

#### 5.1 Gradual deprecation (v2.0)
- [ ] Mark old modules as deprecated with warnings
- [ ] Maintain parallel implementations temporarily
- [ ] Default to new implementation with opt-out flag

#### 5.2 Complete migration (v3.0)
- [ ] Remove `multidms.biophysical` module
- [ ] Remove old optimization code from `multidms.model`
- [ ] Clean up compatibility layers
- [ ] Finalize API

### Option 1: Minimal Breaking Changes (Recommended for v2.0)
```python
# Preserve existing API where possible
model = multidms.Model(
    data=data,
    epistatic_model="Sigmoid",  # Maps to jaxmodels.Sigmoid
    # ... other params
)
model.fit(
    scale_coeff_lasso_shift=1e-5,  # Maps to fusionreg
    # ... other params
)
```

### Option 2: Clean Break (Consider for v3.0)
```python
# New cleaner API
import multidms.jaxmodels as mdms

data_sets = {
    cond: mdms.Data.from_multidms(data, cond) 
    for cond in data.conditions
}
model, trajectory = mdms.fit(
    data_sets,
    reference_condition="wildtype",
    l2reg=1e-4,
    fusionreg=1e-5,
    global_epistasis=mdms.Sigmoid(),
)
```

## Testing Strategy

1. **Regression tests**: Ensure existing notebooks/scripts produce similar results
2. **Performance benchmarks**: Compare fitting speed and memory usage
3. **Numerical validation**: Use `simulation_validation.ipynb` as primary test
4. **API compatibility tests**: Verify old code works with compatibility layer

## Implementation Priority

1. **High Priority**:
   - Core model fitting functionality
   - Data compatibility layer
   - Basic plotting/analysis methods

2. **Medium Priority**:
   - Complete feature parity for all epistasis models
   - ModelCollection integration
   - Comprehensive documentation

3. **Low Priority**:
   - Performance optimizations
   - Additional loss functions
   - Extended visualization options

## Success Criteria

- [ ] All existing notebooks run with minimal modifications
- [ ] Performance improvement of at least 20% on standard datasets
- [ ] Cleaner, more maintainable codebase
- [ ] Improved numerical stability (fewer convergence issues)
- [ ] Clear migration path for existing users

## Risks & Mitigation

1. **Breaking changes**: Provide compatibility layer and clear migration guide
2. **Performance regression**: Benchmark extensively before release
3. **Numerical differences**: Document and justify any changes in results
4. **User confusion**: Extensive documentation and examples

## Timeline Estimate

- Phase 1-2: 2-3 weeks (core functionality)
- Phase 3: 1-2 weeks (analysis/visualization)
- Phase 4: 1 week (documentation/testing)
- Phase 5: Ongoing over 2-3 release cycles

## Related Issues/PRs

- PR #164: Initial jaxmodels implementation
- Consider creating separate tracking issues for each phase

## Questions for Discussion

1. Should we maintain full backward compatibility or allow some breaking changes?
2. What's the preferred timeline for deprecating old modules?
3. Should visualization be separated into its own module?
4. How should we handle the count-based vs functional score loss selection?
5. What level of API similarity should we maintain with the current version?

---

cc: @wsdewitt @jgallowa

**Comment by jgallowa07** [comment]
This is deprecated I'm working on something better.
